op-node/node/config_persistence.go

```diff
@@ -0,0 +1,143 @@
+package node
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+)
+
+type RunningState int
+
+const (
+	StateUnset RunningState = iota
+	StateStarted
+	StateStopped
+)
+
+type persistedState struct {
+	SequencerStarted *bool `json:"sequencerStarted,omitempty"`
+}
+
+type ConfigPersistence interface {
+	SequencerStarted() error
+	SequencerStopped() error
+	SequencerState() (RunningState, error)
+}
+
+var _ ConfigPersistence = (*ActiveConfigPersistence)(nil)
+var _ ConfigPersistence = DisabledConfigPersistence{}
+
+type ActiveConfigPersistence struct {
+	lock sync.Mutex
+	file string
+}
+
+func NewConfigPersistence(file string) *ActiveConfigPersistence {
+	return &ActiveConfigPersistence{file: file}
+}
+
+func (p *ActiveConfigPersistence) SequencerStarted() error {
+	return p.persist(true)
+}
+
+func (p *ActiveConfigPersistence) SequencerStopped() error {
+	return p.persist(false)
+}
+
+// persist writes the new config state to the file as safely as possible.
+// It uses sync to ensure the data is actually persisted to disk and initially writes to a temp file
+// before renaming it into place. On UNIX systems this rename is typically atomic, ensuring the
+// actual file isn't corrupted if IO errors occur during writing.
+func (p *ActiveConfigPersistence) persist(sequencerStarted bool) error {
```

Comment:
Not a bad idea, but in this case I think it has some downsides.
I'm deliberately avoiding having `persistedState` as part of the public interface so the `SequencerState` method can't return a `persistedState` as suggested in the proposed `SequencerStarted` code.  That would mean creating a new `persistedState` in `SequencerStarted` which is problematic if additional settings are persisted in the future because it would silently removed them from the persisted version.  By having the sequencer state be passed as a boolean to `persist` we avoid that and ensure that if a field is added in the future we'll get a compile error to force us to update the code.  There are a few ways we might resolve that in the future depending on what the additional setting is, how frequently it changes and whether it changes at the same time as `SequencerStarted`.

Reply:
Feel free to resolve this comment. @ajsutton